### ISSUE TITLE
Rename hash2, hash2Next, hashUpper, hash2Upper

### ISSUE THREAD

**Issue by holtgrewe** [seqan/seqan#532]
The names of the functions hash2, hash2Next, hashUpper and hash2Upper are not clear enough to get the idea immediately. 

As far as I understand, the function hash2 forces the computation of the hash value even if the sequence is smaller than the actual span of the shape. This also implies, that hash would crash or give a wrong result?

How does the function shapeUpper differs from hash2? In the remarks it is described as a function that computes the hash for a p-gram on an q-gram index, while p<q. So the text is smaller than the span of the shape, as the shape is the property of the q-gram index?

Except of the weak documentation I would suggest changing the interface. Maybe using tags could be helpful. Thus, there is a single hash function whose behavior can be adapted using tags. The following code snippet demonstrates the basic idea.

``` cpp
hash(shape, it, charsLeft, SingleQGram<>());            //-> hash()
hash(shape, it, charsLeft, SingleQGram<Forced<>>());    //-> hash2()
hash(shape, it, charsLeft, AdjacentQGram<>());          //-> hashNext()
hash(shape, it, charsLeft, AdjacentQGram<Forced<>>());  //-> hash2Next()
hash(shape, it, charsLeft, PrefixQGram<>());            //-> hashUpper()
hash(shape, it, charsLeft, PrefixQGram<Forced<>>());    //-> hash2Upper()
```

Changing the interface is not part of the 1.4.1 release, but might be considerable for future releases. However, in the current state at least all hash derivates should link to the others. At the moment this is rather arbitrary.

**Comment by xp3i4** [comment]
I think the interface of hash and hashNext which takes a standard iterator as a parameter makes it difficult to check if  the iterator is out of boundary when you are doing operations in the function like it++ or it--.  Maybe it will be better to use a container instead of a iterator.

**Comment by rrahn** [comment]
We need to reconsider this for SeqAn3. We will adapt the interfaces anyway and need to reflect where all these intrfaces are really used.

**Comment by smehringer** [comment]
Maybe open a meta issue for the qgram index which involves hashing and reference this issue.

**Comment by rrahn** [comment]
Yes, but while I was going through the list, I though it would be better to do it with thre new development famework and we have to discuss the place where this shall live. More on this on Monday 😏
